### ISSUE TITLE
Shared tests

### ISSUE THREAD

**Issue by mlrawlings** [webcomponents/custom-elements-everywhere#7]
Going through the tests, it seemed like there were some (small) disparities between some of the libraries.  This [dojo test in particular](https://github.com/webcomponents/custom-elements-everywhere/blob/1fcc924a0dea84a0958f1ab8d84a52c7ac3d79d6/libraries/dojo2/src/component-tests.ts#L85-L91) caught my eye.

If the tests were shared, it would give a level of consistency across the tests.   It would also make it easier to add new tests/fix tests (if necessary) and to add new libraries to the repo.  It might even be useful to export the tests so that libraries could add them to their own test suites.

I'm thinking providing a light adaptor for each framework would be enough. 

Take a look at this example for Preact:
https://gist.github.com/mlrawlings/1a14ddf446ca9787da7e378bc80f9a0c
_conceptually, it should work, but I haven't actually run the code_

Thoughts?

**Comment by tbranyen** [comment]
This was my sentiment as well. I was going to add diffHTML tests, but it feels like I could just change the test to make it work the way I think it should and not necessarily what would grant us actual consistency.

**Comment by robdodson** [comment]
Thanks for doing all of this work @mlrawlings. I need to spend some more time going through it as there are a few spots I don't quite understand. My ES2015 skills leave much to be desired 😅 

The thought of centralizing all of the tests has definitely occurred to me, but I decided against it because I was concerned about the number of abstractions involved. I'm worried that building a meta testing platform and adapters per framework makes it harder for new folks to contribute to the project. At the moment you can just start writing tests in the manner you're most familiar with, and so long as you understand the intent it's pretty easy for you to contribute to the project (i hope at least!)

> I was going to add diffHTML tests, but it feels like I could just change the test to make it work the way I think it should and not necessarily what would grant us actual consistency.

I think if you're doing that then I've failed as a test reviewer 😛  I think in the case of the cited DOJO tests, I need to go back and give those a second look. I should also more thoroughly explain what each test aims to achieve. Maybe in the readme or a wiki.

**Comment by mlrawlings** [comment]
Specifically with those Dojo tests, the [lightdom test](https://github.com/webcomponents/custom-elements-everywhere/blob/1fcc924a0dea84a0958f1ab8d84a52c7ac3d79d6/libraries/dojo2/src/component-tests.ts#L77-L83) doesn't make any assertions about the lightdom that was passed in.  And it's using a re-render component (the test description doesn't really match this, but this is at least consistent across the tests).  Other tests, like the preact test do assert the original lightdom, wait for the promise to resolve, the dom to update, and then assert the new lightdom.

The [hide/show test](https://github.com/webcomponents/custom-elements-everywhere/blob/1fcc924a0dea84a0958f1ab8d84a52c7ac3d79d6/libraries/dojo2/src/component-tests.ts#L85-L91) doesn't call the toggle method, so it isn't testing anything more than the previous test.

-----

I'm planning on making a PR to add Marko later this week probably.  If you're mostly okay with what you see in that shared tests file, I'll use that for the Marko tests and then we could see about adapting the other tests to use it as well.  And @tbranyen, perhaps you could use that as the starting point for diffHTML.

**Comment by tbranyen** [comment]
Sounds good to me @mlrawlings thanks for taking the initiative!

**Comment by robdodson** [comment]
Some running notes as I read through this...

```
let NoChildren = getComponent('NoChildren',`
  <ce-without-children></ce-without-children>
`);
```
I'm not sure I understand the point of the second argument to `getComponent`. It doesn't seem to be used for anything except to log an error. Is the intention that the 2nd argument would go inside of the Component or is it really just for logging purposes?

[mountAndGetComponent](https://gist.github.com/mlrawlings/1a14ddf446ca9787da7e378bc80f9a0c#file-preact-tests-js-L12) doesn't seem to actually be used?

Is [this](https://gist.github.com/mlrawlings/1a14ddf446ca9787da7e378bc80f9a0c#file-run-tests-js-L125) supposed to be `call()`?

Looking at the more advanced tests, [like this one](https://gist.github.com/mlrawlings/1a14ddf446ca9787da7e378bc80f9a0c#file-run-tests-js-L109-L117), I'm not sure if I understand the implementation. Is the idea that the template string would go inside of your component somehow? Because each library has its own syntax for binding data into a component I'm not sure how this would work.

**Comment by mlrawlings** [comment]
@robdodson 
> Is the intention that the 2nd argument would go inside of the Component or is it really just for logging purposes? [...]  Is the idea that the template string would go inside of your component somehow? Because each library has its own syntax for binding data into a component I'm not sure how this would work.

Yes, it is just for logging purposes.  The idea is that if you're missing a component it will log the component name that is missing plus a concise "framework agnostic" representation of what the component should look like.

Though, in reality, it would probably be easier to look at another one of the frameworks, but I thought it might prove a bit helpful.

Perhaps adding text to the effect of _"The component should look a little something like this, but in your own framework's syntax"_ would help.  Or maybe it's just confusing and should be removed.

> mountAndGetComponent doesn't seem to actually be used?
> Is this supposed to be call()?

There were some issues with the code I posted.  I had renamed some things and not updated references.  As I said, I never actually ran the code.  Here is a [working version](https://github.com/mlrawlings/custom-elements-everywhere/blob/master/libraries/marko/src/run-tests.js).

I'll submit a PR for this next week once we land some changes in Marko to [better support custom elements](https://github.com/marko-js/marko/compare/custom-elements).
